File: frontend/src/api/openclaw-config.ts
import { customFetch } from "./mutator";

export type ConfigTreeNode = {
  name: string;
  type: "dir" | "file";
  children: ConfigTreeNode[] | null;
};

export type ConfigTreeResponse = {
  root: string;
  tree: ConfigTreeNode;
};

type ApiResponse<T> = { data: T; status: number; headers: Headers };

export async function getConfigTree(): Promise<ConfigTreeResponse> {
  const response = await customFetch<ApiResponse<ConfigTreeResponse>>(
    "/api/v1/openclaw/config-tree",
    { method: "GET" },
  );
  return response.data;
}

File: backend/app/schemas/openclaw_config.py
"""Schemas for OpenClaw config directory tree API."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ConfigTreeNode(BaseModel):
    """A node in the OpenClaw config directory tree."""

    name: str = Field(description="File or directory name")
    type: Literal["dir", "file"] = Field(description="Whether this is a directory or file")
    children: list["ConfigTreeNode"] | None = Field(
        default=None,
        description="Child nodes for directories; null for files",
    )


ConfigTreeNode.model_rebuild()


class ConfigTreeResponse(BaseModel):
    """Response payload for the config directory tree."""

    root: str = Field(description="Absolute path of the root directory")
    tree: ConfigTreeNode = Field(description="Root node of the directory tree")
